protobuf/source_reflection.go

```diff
@@ -0,0 +1,71 @@
+package protobuf
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/grpcreflect"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	rpb "google.golang.org/grpc/reflection/grpc_reflection_v1alpha"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/resolver/manual"
+)
+
+var (
+	// ErrorCouldNotDialReflectionServer is an error for when the reflection server could not be reached.
+	ErrorCouldNotDialReflectionServer = errors.New("could not reach reflection server")
+)
+
+// ReflectionArgs are args for constructing a DescriptorProvider that reaches out to a reflection server.
+type ReflectionArgs struct {
+	Caller  string
+	Service string
+	Peers   []string
+	Timeout time.Duration
+}
+
+// NewDescriptorProviderReflection returns a DescriptorProvider DescriptorProvider that reaches out to
+// a reflection server to access filedescriptors.
+func NewDescriptorProviderReflection(args ReflectionArgs) (DescriptorProvider, error) {
+	metadataContext := metadata.NewOutgoingContext(context.Background(),
+		map[string][]string{
+			"rpc-caller":   []string{args.Caller},
+			"rpc-service":  []string{args.Service},
+			"rpc-encoding": []string{"proto"},
+		})
+	r, deregisterScheme := manual.GenerateAndRegisterManualResolver()
+	defer deregisterScheme()
```

Comment:
nit: should we rename `deregisterScheme` to `unregister` or `cleanup`?

Reply:
I actually remember reading through https://english.stackexchange.com/questions/25931/unregister-vs-deregister when I was deciding on `de-` vs `un-` register. Thought this fit best.

Reply:
Nice, thanks for the reference! Seems reasonable to me.